Add create plugin update action

jackw opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/plugin-tools/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
This PR adds an action that automates `create-plugin update` commands to help plugin devs keep their create-plugin related tooling, dependencies, and go sdk up to date.

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Related grafana/plugin-tools/issues/818

**Special notes for your reviewer**:
I'm not sure if there any glaring security issue related to passing the GH token with raised privileges to this composite workflow. For now I've added it but if there's concern from reviewers we can move it out to the example usage workflow so the plugin's workflow owns the token and open PR step.

jackw committed:
feat(create-plugin-update): initial commit of action to run update cmd

jackw committed:
fix(cp-update): fix typo in bash script

jackw committed:
Wip

jackw committed:
Wip

jackw committed:
Wip

jackw committed:
Wip

jackw committed:
Wip

jackw committed:
docs(cp-update): clean up readme

jackw committed:
Wip

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/plugin-actions?pullRequest=38) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/plugin-actions?pullRequest=38) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/plugin-actions?pullRequest=38) it.</sub>

sunker commented on create-plugin-update/README.md:
```diff
@@ -0,0 +1,54 @@
+
+**Error: GitHub Actions is not permitted to create or approve pull requests.**
+
+If you're seeing this error it means the GH token passes to the action doesn't have the necessary privileges to create the pull request. To resolve this you can either:
```
```suggestion
If you're seeing this error it means the GH token passed to the action doesn't have the necessary privileges to create the pull request. To resolve this you can either:
```

jackw replied:
Ah I resolved this manually whilst making other changes. 😞

sunker commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,85 @@
+  node-version:
+    description: "Version of node"
+    required: false
+    default: "20"
```
Maybe default to `lts/*` rather than a specific version?

sunker reviewed (approved):
Someone else can probably comment on the security around tokens. But this is a really nice addition! 🚀

andresmgot commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,85 @@
+        delete-branch: true
+        commit-message: "chore: update configuration to create-plugin ${{ env.LATEST_VERSION }}"
+        title: "Chore: Bump @grafana/create-plugin configuration to ${{ env.LATEST_VERSION }}"
+        author: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
```
is this author accurate? (Shouldn't that be inferred from the token?)

jackw replied:
I don't know if it's accurate but we cannot push unless the use name and email are set in the git config. Using the github bot seemed like a reasonable default choice. We could offer it as an option if devs would like to customise it?

andresmgot commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,85 @@
+      shell: bash
+
+    - name: Create PR
+      uses: peter-evans/create-pull-request@v7
```
if the security concern is passing the token to a thirdparty, you can always do it manually with the gh cli `gh pr create ...` ([docs](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/using-github-cli-in-workflows))

academo replied:
+1 on using the github cli for this instead of a third party action.

andresmgot reviewed (commented):
Some thoughts here

academo commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,85 @@
+name: "Grafana Build Plugin"
```
Is this name and description correct?

jackw replied:
Nope, copy n pasta mistake on my part. Should be fixed now.

academo commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,85 @@
+name: "Grafana Build Plugin"
+description: "Builds a Grafana plugin"
+
+outputs:
```
are any of these outputs relevant?

jackw replied:
Nope, copy n pasta mistake on my part.

jackw committed:
feat(create-plugin-update): replace 3rd party create pr action with gh cli

jackw committed:
feat(create-plugin-update): get credentials working and tidy up

academo commented on create-plugin-update/README.md:
````diff
@@ -0,0 +1,54 @@
+- Go to https://github.com/USER_NAME/REPO_NAME/settings/actions and check **Allow GitHub Actions to create and approve pull requests** then add the following to your workflow to elevate the token permissions:
+
+  ```yaml
+  permissions:
````
should this be part of the example we provide to use the action?

jackw replied:
I've moved it up into the example rather than keep it in a separate section. Let me know what you think.

academo commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,93 @@
+name: "Grafana Create Plugin Update"
+description: "Runs create-plugin update within a Grafana plugin"
```
Could you change this to specify this actually creates a PR? 

`Runs create-plugin update within a Grafana plugin and opens a PR with the changes`

jackw replied:
Have updated as suggested. 👍

jackw committed:
docs(create-plugin-update): update readme and action description

jackw committed:
feat(create-plugin-update): support running from manual workflow runs using a different branch

andresmgot commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,94 @@
+        git checkout -b update-grafana-create-plugin
+        git add .
+        git commit -m "chore: update configuration to create-plugin ${{ env.LATEST_VERSION }}"
+        git push --force https://x-access-token:${{ inputs.token }}@github.com/${{ github.repository }} update-grafana-create-plugin
```
maybe not forcing this? In case there are changes already in the remote branch that needs care

jackw replied:
Hmm... you're right this needs a rethink. Right now it always works with the same branch name and the action I was using was force pushing to keep the single branch updated. This has worked well for my little plugin for the last 5 months or so but I see your point - it would be rather painful if a dev spent a load of time fixing up the PR then the next morning all their changes are gone and have to wrangle the branch.

Do you think each run that has an update should open a separate PR against a version specific branch (e.g. `update-grafana-create-plugin-${{ env.LATEST_VERSION }}`)? Like this the action would only ever push one commit to a specific branch and there wouldn't need to be force involved ever. Plugin devs could set it to run every month or so to keep noise down and they would be free to tackle any issues without the action causing them bother.

jackw replied:
There's a strong chance that right now a dev will want to do some manual intervention as the update command updates all the grafana npm packages to latest which isn't always ideal.

andresmgot replied:
> Do you think each run that has an update should open a separate PR against a version specific branch (e.g. `update-grafana-create-plugin-${{ env.LATEST_VERSION }}`)?

That sounds good (plus not forcing, in case the same GH action is executed twice).

jackw replied:
Added in https://github.com/grafana/plugin-actions/pull/38/commits/cabd6b49de3d2ad39f50bc5ecb692bc6f9e025b4

jackw commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,85 @@
+        ${{ github.action_path }}/pm.sh update
+        ${{ github.action_path }}/pm.sh install
+      shell: bash
+
```
There will always be changes as the .cprc file keeps a record of the [version of create-plugin](https://github.com/jackw/heywesty-trafficlight-panel/pull/41/files#diff-d1ad74bd29d5e81e157a0f0098623cb8c13603cfb2e06efb5c2569cf0eb9db26) that was run.

andresmgot reviewed (approved):
LGTM, just one comment (not sure if this is ready)

jackw committed:
feat(create-plugin-update): use version specific branches and dont force push

jackw committed:
docs(create-plugin-update): update readme detailing what the action does and  example workflow cron

jackw commented:
Have now implemented individual branches per version. A successful run example can [be found here](https://github.com/jackw/heywesty-trafficlight-panel/pull/42).

This means that the action will now open a separate PR (and associated branch) for each new version of create-plugin. This can create noise so I've updated the readme example to run once a month.

leventebalogh commented on create-plugin-update/README.md:
```diff
@@ -0,0 +1,47 @@
+# Create plugin update action
+
+This GitHub Action automates the process of running `create-plugin update` within your plugins repository. It checks the current create-plugin version against latest and if there is a newer version available will create a branch, run create-plugin update, update node lock file, and open a PR with the changes.
```
```suggestion
This GitHub Action automates the process of running `create-plugin update` within your plugins repository. It checks the current create-plugin version against latest and if there is a newer version available it will create a branch, run create-plugin update, update the node lock file, and open a PR with the changes.
```

leventebalogh commented on create-plugin-update/action.yml:
```diff
@@ -0,0 +1,94 @@
+        git config --local user.name "github-actions[bot]"
+        git config --local user.email "github-actions[bot]@users.noreply.github.com"
+        git checkout -b update-grafana-create-plugin-${{ env.LATEST_VERSION }}
+        git add .
```
Are we sure we don't want to stage any deletions (`git add .`) ?

leventebalogh replied:
(My bad, `git add .` actually stages deletions, too.)

leventebalogh commented on create-plugin-update/pm.sh:
```diff
@@ -0,0 +1,50 @@
+
+# Check if command argument is provided
+if [ "$1" = "" ]; then
+	echo "Please provide a command to run."
```
Nit, maybe.
```suggestion
	echo "Please provide a command to run. Available commands: install, update"
```

leventebalogh reviewed (approved):
LGTM 👍

andresmgot reviewed (approved):
LGTM

jackw committed:
Apply suggestions from code review

Co-authored-by: Levente Balogh <balogh.levente.hu@gmail.com>
